File: jeph2sworm-core/jeph2sworm/llm/openai_provider.py
"""OpenAI Provider - GPT-4o, o1, o3 models."""

from __future__ import annotations

from typing import List, Optional

import structlog

from jeph2sworm.llm.base_provider import BaseProvider, LLMMessage, LLMResponse

logger = structlog.get_logger()

MODELS = [
    "gpt-4o",
    "gpt-4o-mini",
    "gpt-4-turbo",
    "o1",
    "o1-mini",
    "o1-preview",
    "o3-mini",
]

TASK_BEST = {
    "coding": "gpt-4o",
    "planning": "o1",
    "reasoning": "o1",
    "design": "gpt-4o",
    "testing": "gpt-4o-mini",
    "general": "gpt-4o",
}


class OpenAIProvider(BaseProvider):
    provider_name = "openai"
    supports_function_calling = True
    supports_vision = True
    max_context_window = 128_000

    def available_models(self) -> List[str]:
        return MODELS

    def best_model_for(self, task: str) -> str:
        return TASK_BEST.get(task, "gpt-4o")

    async def complete(
        self,
        messages: List[LLMMessage],
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs,
    ) -> LLMResponse:
        import litellm

        response = await litellm.acompletion(
            model=f"openai/{model}" if not model.startswith("openai/") else model,
            messages=[m.model_dump() for m in messages],
            temperature=temperature,
            max_tokens=max_tokens,
            api_key=self.api_key,
            **kwargs,
        )

        choice = response.choices[0]
        usage = dict(response.usage) if response.usage else {}

        return LLMResponse(
            content=choice.message.content or "",
            model=model,
            provider=self.provider_name,
            usage=usage,
            finish_reason=choice.finish_reason or "stop",
        )

File: jeph2sworm-vscode/src/providers/openai-provider.ts
/**
 * OpenAI Provider — proxy to GPT-4o / GPT-4-turbo / o1 models.
 */

import { BaseProvider, LLMMessage, LLMResponse, ProviderConfig } from './base-provider';

export class OpenAIProvider extends BaseProvider {
  readonly name = 'openai';
  readonly models = ['gpt-4o', 'gpt-4-turbo', 'gpt-4o-mini', 'o1', 'o1-mini'];

  constructor(config: ProviderConfig) {
    super({ baseUrl: 'https://api.openai.com/v1', ...config });
  }

  async complete(messages: LLMMessage[], model?: string): Promise<LLMResponse> {
    const chosen = model ?? this.config.defaultModel ?? 'gpt-4o';
    const res = await fetch(`${this.config.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({ model: chosen, messages }),
    });
    const data = await res.json() as any;
    return {
      content: data.choices?.[0]?.message?.content ?? '',
      model: chosen,
      tokensUsed: data.usage?.total_tokens ?? 0,
      finishReason: data.choices?.[0]?.finish_reason ?? 'unknown',
    };
  }

  bestModelFor(task: 'coding' | 'planning' | 'design' | 'quick'): string {
    switch (task) {
      case 'coding':
        return 'gpt-4o';
      case 'planning':
        return 'o1';
      case 'design':
        return 'gpt-4o';
      case 'quick':
        return 'gpt-4o-mini';
    }
  }
}
